components/higher-order/with-focus-contain/index.js

```diff
@@ -0,0 +1,57 @@
+/**
+ * WordPress dependencies
+ */
+import { Component, createRef } from '@wordpress/element';
+import { keycodes } from '@wordpress/utils';
+import { focus } from '@wordpress/dom';
+
+const { TAB } = keycodes;
+
+const withFocusContain = ( WrappedComponent ) => {
+	return class extends Component {
+		constructor() {
+			super( ...arguments );
+
+			this.focusContainRef = createRef();
+			this.handleTabBehaviour = this.handleTabBehaviour.bind( this );
+		}
+
+		handleTabBehaviour( event ) {
+			if ( event.keyCode !== TAB ) {
+				return;
+			}
+
+			const tabbables = focus.tabbable.find( this.focusContainRef.current );
+			if ( ! tabbables.length ) {
+				return;
+			}
+			const firstTabbable = tabbables[ 0 ];
+			const lastTabbable = tabbables[ tabbables.length - 1 ];
+
+			if ( event.shiftKey && event.target === firstTabbable ) {
+				event.preventDefault();
+				lastTabbable.focus();
+			} else if ( ! event.shiftKey && event.target === lastTabbable ) {
+				event.preventDefault();
+				firstTabbable.focus();
+			}
+		}
+
+		render() {
+			// Disable reason: this component is non-interactive, but must capture
+			// events from the wrapped component to determine when the Tab key is used.
+			/* eslint-disable jsx-a11y/no-static-element-interactions */
+			return (
+				<div
+					onKeyDown={ this.handleTabBehaviour }
+					ref={ this.focusContainRef }
+				>
+					<WrappedComponent { ...this.props } />
+				</div>
+			);
+			/* eslint-enable jsx-a11y/no-static-element-interactions */
+		}
+	};
+};
```

Comment:
I'm pretty sure a similar mechanism is also implemented elsewhere. Maybe NavigableContainer? Should we consolidate?

Reply:
> Worth also considering the name should clarify this is something different from what other components do, for example NavigableContainer.

🙂 Yes and no. The mechanism in NavigableContainer is used also for elements with tabindex="-1" like menu items.